cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+
+use std::vec::Vec;
+
+/// The peephole optimiser's state.
+pub struct Peephole {}
+
+impl Peephole {
+    /// Create a new peephole optimiser state.
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    /// Clear the state of the peephole optimiser.
+    pub fn clear(&mut self) {}
+
+    /// Run the peephole optimiser.
+    pub fn run(&mut self, isa: &TargetIsa, func: &mut Function, domtree: &DominatorTree) {
+        let mut ctx = Context {
+            cur: EncCursor::new(func, isa),
+            reginfo: isa.register_info(),
+            domtree,
+        };
+        ctx.run()
+    }
+}
+
+struct Context<'a> {
+    // Current instruction as well as reference to function and ISA.
+    cur: EncCursor<'a>,
+
+    // Cached ISA information.
+    // We save it here to avoid frequent virtual function calls on the
+    // `TargetIsa` trait object.
+    reginfo: RegInfo,
+
+    // References to contextual data structures we need.
+    domtree: &'a DominatorTree,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        self.cleanup();
+    }
+}
+
+// ======== A description of the redundant-fill-removal algorithm ========
+//
+// The simplest incarnation simply detects fills from the same stack slot into
+// the same register, and removes them:
+//
+//  v1 = fill v0
+//  ...
+//  v2 = fill v0
+//
+// The second fill is redundant provided that (1) v1 and v2 have the same
+// register, (2) that "..." overwrites neither that register nor v0's stack
+// slot, and (3) that the second fill is dominated by the first.  (3) is
+// implied since we're doing this only inside EBBs, but we have to check (1)
+// and (2).
+//
+// In this example, the instruction processing loop below creates an
+// environment entry on encountering "v1 = fill v0", and uses that entry to
+// determine that "v2 = fill v0" is redundant.
+//
+// The loop must also, even in the simplest version, check all instructions
+// for writes to registers and stack slots, and invalidate environment entries
+// accordingly.
+//
+// === First refinement -- implemented
+//
+// This is to observe that spills also create equivalences between registers
+// and stack slot values, which in turn increases our changes of proving later
+// fills redundant.
+//
+// === Second refinement -- implemented
+//
+// For a fill, even if we can't find an exact match in the environment,
+// finding a load of the same spill slot into a different register is still a
+// win, since we can convert the fill into a reg-reg copy.
+//
+// === Other tricks that don't seem to help
+//
+// I tried tracking the effects of InstructionData::Copy in the environment,
+// for the case when both source and destination are registers.  But that
+// didn't cause any more fills to be removed or converted to copies.  I'm not
+// sure why.
+//
+// It's also possible for a spill to be detected as redundant, if such a
+// binding already appears in the environment.  I implemented this, but it
```

Comment:
There really should not be many redundant spills, if any, given how the register allocator assigns locations.  Once a value is flagged as spilled, it is going to be spilled everywhere and is spilled upon creation; since values are immutable, there's never any reason to spill a value again while it is live.  The live range splitter may change this calculus, but should have its own optimizations to avoid redundant spills.

Reply:
Yes.  I think I did observe one or perhaps two in our "standard 6" set of perf benchmarks.